src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -31,14 +40,100 @@ internal ModelOperationsCatalog(IHostEnvironment env)
         /// <param name="stream">A writeable, seekable stream to save to.</param>
-        public void Save(ITransformer model, Stream stream) => model.SaveTo(_env, stream);
+        public void Save<TSource>(IDataLoader<TSource> model, Stream stream)
+        {
+            using (var rep = RepositoryWriter.CreateNew(stream))
+            {
+                ModelSaveContext.SaveModel(rep, model, "Model");
+                SaveInputSchema(model.GetOutputSchema(), rep);
+                rep.Commit();
+            }
+        }
+
+        /// <summary>
+        /// Save a transformer model and the loader used to create its input data to the stream.
+        /// </summary>
+        /// <param name="loader">The loader that was used to create data to train the model</param>
+        /// <param name="model">The trained model to be saved</param>
+        /// <param name="stream">A writeable, seekable stream to save to.</param>
+        public void Save<TSource>(IDataLoader<TSource> loader, ITransformer model, Stream stream) =>
+            Save(new CompositeDataLoader<TSource, ITransformer>(loader, new TransformerChain<ITransformer>(model)), stream);
+
+        /// <summary>
+        /// Save a transformer model and the schema of the data that was used to train it to the stream.
+        /// </summary>
+        /// <param name="inputSchema">The schema of the input to the transformer.</param>
+        /// <param name="model">The trained model to be saved.</param>
+        /// <param name="stream">A writeable, seekable stream to save to.</param>
+        public void Save(DataViewSchema inputSchema, ITransformer model, Stream stream)
+        {
+            using (var rep = RepositoryWriter.CreateNew(stream))
+            {
+                ModelSaveContext.SaveModel(rep, model, TransformerDirectory);
+                SaveInputSchema(inputSchema, rep);
+                rep.Commit();
+            }
+        }
+
+        private void SaveInputSchema(DataViewSchema inputSchema, RepositoryWriter rep)
+        {
+            using (var ch = _env.Start("Saving Schema"))
+            {
+                var entry = rep.CreateEntry(SchemaEntryName);
+                var saver = new BinarySaver(_env, new BinarySaver.Arguments { Silent = true });
+                DataSaverUtils.SaveDataView(ch, saver, new EmptyDataView(_env, inputSchema), entry.Stream, keepHidden: true);
+            }
+        }
 
         /// <summary>
-        /// Load the model from the stream.
+        /// Load the model and its input schema from the stream.
         /// </summary>
         /// <param name="stream">A readable, seekable stream to load from.</param>
+        /// <param name="inputSchema">Will contain the input schema for the model.</param>
         /// <returns>The loaded model.</returns>
-        public ITransformer Load(Stream stream) => TransformerChain.LoadFrom(_env, stream);
+        public ITransformer Load(Stream stream, out DataViewSchema inputSchema)
```

Comment:
>DataViewSchema inputSchema [](start = 52, length = 26)

Sometimes this will not be present. Should we clarify that this can be null in some cases?